File: web/src/api/scheduling.ts
// 编制域登记面、队列与协调者生命周期 wire 的唯一 web fetch 面。
// 与 internal/proto/scheduling.go 对齐；testdata fixture 和 contract.test.ts 钉住
// 线格式。本文件不管理业务状态，也不解析、拼接或替换服务端返回的命令。
import { postJSON, putJSON, request } from './client'

export interface CarrierView {
  name: string
  machine: string
  cli: string
  home_dir: string
  model?: string
  credential: string
  max_concurrency?: number
  status?: CarrierStatus
  last_error?: string
  version: number
}

export type CarrierStatus = 'pending' | 'online' | 'quota' | 'unreachable'
export type ProbeKind = 'empty' | 'logged_in' | 'occupied'
export type WakeOutcome = 'ready' | 'need_login' | 'quota' | 'unreachable'

export interface SquadView {
  name: string
  role: string
  members: SquadMember[]
  version: number
}

export interface SquadMember {
  carrier: string
  max_concurrency?: number
}

export interface SquadsResp {
  carriers: CarrierView[]
  squads: SquadView[]
}

export interface CarrierInput {
  name?: string
  machine: string
  cli: string
  home_dir: string
  model?: string
  credential: string
  max_concurrency?: number
}

export interface SquadInput {
  name?: string
  role: string
  members: SquadMember[]
}

export interface SquadPutResp {
  name: string
  version: number
}

export interface QueueEntry {
  kind: string
  id: string
  card: string
  node?: string
  squad: string
  target?: string
  executor?: string
  model?: string
  priority?: string
  ready: boolean
  actor: string
  seq: number
  position: number
}

export interface QueueResp {
  queue: QueueEntry[]
}

export interface CoordinatorLaunchResp {
  woke: boolean
  session_id?: string
  rebuilt: boolean
  escalated: boolean
  output?: string
}

/** POST attach=true 的服务端定位回执；machine 空串仍须保留。 */
export interface CoordinatorAttachInfo {
  machine: string
  dir: string
  command: string
}

/** GET coordinator 的状态；未绑定时 attach 明确为 null。 */
export interface CoordinatorStatus {
  bound: boolean
  attach_active: boolean
  attach: CoordinatorAttachInfo | null
}

/** POST attach=false 的服务端成功回执。 */
export interface CoordinatorAttachReleaseResp {
  ok: boolean
}

/** 获取载体与小队快照；空数组和服务端错误均原样交给调用方。 */
export const getSquads = () => request<SquadsResp>('/api/squads')

/** 获取瞬时调度队列；本函数不把空队列改写为缺席值。 */
export const getQueue = () => request<QueueResp>('/api/queue')

// 服务端用 max_concurrency 缺席表达「不限」；保护 API 边界不把调用方传入的
// 0 序列化成另一种语义，尤其是表单从数字输入转换而来的零值。
function omitZeroConcurrency<T extends CarrierInput>(input: T): T {
  if (input.max_concurrency !== 0) return input
  const copy = { ...input }
  delete copy.max_concurrency
  return copy
}

/** 按载体名和 CAS 版本更新载体；expect=0 表示新建。 */
export const putCarrier = (name: string, expect: number, input: CarrierInput) =>
  putJSON<SquadPutResp>(
    `/api/squads/carriers/${encodeURIComponent(name)}?expect=${expect}`,
    omitZeroConcurrency(input),
  )

// 服务端用成员 max_concurrency 缺席表达「不限」；仅将调用方已经给出的 0
// 投影为缺席，不能把非法文本在 API 层静默变成不限，页面负责输入校验。
function omitZeroSquadConcurrency(input: SquadInput): SquadInput {
  return {
    ...input,
    members: input.members.map((member) => {
      if (member.max_concurrency !== 0) return member
      const copy = { ...member }
      delete copy.max_concurrency
      return copy
    }),
  }
}

/** 按小队名和 CAS 版本更新小队；expect=0 表示新建，成员政策缺席表示不限。 */
export const putSquad = (name: string, expect: number, input: SquadInput) =>
  putJSON<SquadPutResp>(
    `/api/squads/squads/${encodeURIComponent(name)}?expect=${expect}`,
    omitZeroSquadConcurrency(input),
  )

/** 探测目标机路径；machine 走 query，不进 body。本函数不改写 kind。 */
export const probeHome = (input: HomeProbeReq) => {
  const q = input.machine ? `?machine=${encodeURIComponent(input.machine)}` : ''
  const body: HomeProbeReq = { cli: input.cli, path: input.path }
  if (input.credential) body.credential = input.credential
  return postJSON<HomeProbeResp>(`/api/host/probe${q}`, body)
}

/** 本机/目标机有时限唤起；machine 走 query。控制台不应直接调，检测编排才用。 */
export const wakeHome = (input: HomeWakeReq) => {
  const q = input.machine ? `?machine=${encodeURIComponent(input.machine)}` : ''
  const body: HomeWakeReq = { cli: input.cli, home_dir: input.home_dir }
  if (input.credential) body.credential = input.credential
  if (input.model) body.model = input.model
  return postJSON<HomeWakeResp>(`/api/host/wake${q}`, body)
}

/** 对已登记载体做一次检测写状态。 */
export const detectCarrier = (name: string) =>
  postJSON<CarrierDetectResp>(
    `/api/squads/carriers/${encodeURIComponent(name)}/detect`,
    {},
  )

/** 取服务端生成的运行命令；调用方只复制，不拼接。 */
export const getCarrierRunCommand = (name: string) =>
  request<CarrierRunCommandResp>(
    `/api/squads/carriers/${encodeURIComponent(name)}/run-command`,
  )

/** 四态英文键 → 用户可见中文名；与 scheduling.CarrierStatus.Label 同一份词表。 */
export const CARRIER_STATUS_LABEL: Record<CarrierStatus, string> = {
  pending: '未上线',
  online: '已上线',
  quota: '限额中',
  unreachable: '不可达',
}

/** 登记弹窗默认 HOME 串；空名字返回空串。与 scheduling.DefaultHomeDir 同一格式。 */
export function defaultHomeDir(name: string): string {
  const trimmed = name.trim()
  return trimmed ? `~/.handoff/home/${trimmed}` : ''
}

export interface HomeProbeReq {
  cli: string
  path: string
  credential?: string
  machine?: string
}

export interface HomeProbeResp {
  kind: ProbeKind
  detail?: string
}

export interface HomeWakeReq {
  cli: string
  home_dir: string
  credential?: string
  model?: string
  machine?: string
}

export interface HomeWakeResp {
  outcome: WakeOutcome
  detail?: string
}

export interface CarrierDetectResp {
  name: string
  status: CarrierStatus
  last_error?: string
  version: number
}

export interface CarrierRunCommandResp {
  command: string
}

export type CoordinatorLaunchSource = 'manual' | 'card_create'

/**
 * 拉起协调者回合；省略 source 时发送空对象以保留服务端 manual 默认值，
 * card_create 只用于开卡即绑审计。返回的回合结果由服务端产生。
 */
export const launchCoordinator = (cardId: string, source?: CoordinatorLaunchSource) =>
  postJSON<CoordinatorLaunchResp>(
    `/api/cards/${encodeURIComponent(cardId)}/coordinator/launch`,
    source === undefined ? {} : { source },
  )

/** 参数：完整 cardId；返回：GET coordinator wire；路径段必须 encodeURIComponent。 */
export const getCoordinatorStatus = (cardId: string) =>
  request<CoordinatorStatus>(`/api/cards/${encodeURIComponent(cardId)}/coordinator`)

/** 参数：完整 cardId 与 workdir；返回：AttachInfo；不改写服务端三元组。 */
export const attachCoordinator = (cardId: string, workdir: string) =>
  postJSON<CoordinatorAttachInfo>(
    `/api/cards/${encodeURIComponent(cardId)}/attach`,
    { active: true, workdir },
  )

/** 参数：完整 cardId；返回：{ok:true}；路径段必须 encodeURIComponent。 */
export const releaseCoordinator = (cardId: string) =>
  postJSON<CoordinatorAttachReleaseResp>(
    `/api/cards/${encodeURIComponent(cardId)}/attach`,
    { active: false },
  )

File: internal/proto/scheduling.go
// 编制域 wire 面（B156.3 K3）：squads/queue 端点与协调者拉起响应的请求/响应
// DTO。字段名与 web/src/api/scheduling.ts 镜像一字不差；线格式由
// TestContractFixtures 逐字节钉住（改字段先看红，再 -update 显式刷新）。
//
// 边界：proto 不 import 编制域——View 类型是独立投影，scheduling.Carrier/Squad
// → View 的手写投影在 agentd（schedapi.go），该边界由其 wire 往返测试锁。
package proto

// CarrierView 是载体登记行（GET /api/squads 返回元素）。Version 是 registry
// 行版本，CAS 编辑回路（GET 取版 → PUT ?expect=）的唯一权威来源。
type CarrierView struct {
	Name           string `json:"name"`
	Machine        string `json:"machine"`
	CLI            string `json:"cli"`
	HomeDir        string `json:"home_dir"`
	Model          string `json:"model,omitempty"`
	Credential     string `json:"credential"`
	MaxConcurrency int    `json:"max_concurrency,omitempty"`
	Healthy        bool   `json:"healthy"`
	Status         string `json:"status,omitempty"`     // B293：四态英文键；空=尚未回填
	LastError      string `json:"last_error,omitempty"` // 最近一次检测说明；不参与准入
	Version        int    `json:"version"`
}

// SquadView 是小队登记行。MaxConcurrency=0（不限）以键缺席表达。
type SquadView struct {
	Name           string   `json:"name"`
	Role           string   `json:"role"`
	Members        []string `json:"members"`
	MaxConcurrency int      `json:"max_concurrency,omitempty"`
	Version        int      `json:"version"`
}

// SquadsResp 是 GET /api/squads 的响应体。空库时两个字段是空数组而非 null
// ——「什么都没配」是合法态，前端据此渲染引导文案而不是报错。
type SquadsResp struct {
	Carriers []CarrierView `json:"carriers"`
	Squads   []SquadView   `json:"squads"`
}

// CarrierInput 是 PUT 载体的请求体。刻意不含 status / last_error / healthy：
// 状态只由创建、home_dir 变更与检测写入，登记请求不得设置（B293 废止 PutCarrier 翻真）。
type CarrierInput struct {
	Name           string `json:"name,omitempty"`
	Machine        string `json:"machine"`
	CLI            string `json:"cli"`
	HomeDir        string `json:"home_dir"`
	Model          string `json:"model,omitempty"`
	Credential     string `json:"credential"`
	MaxConcurrency int    `json:"max_concurrency,omitempty"`
}

// SquadInput 是 PUT 小队的请求体。Members 为空合法（岔口四：先建空队再补成员）。
type SquadInput struct {
	Name           string   `json:"name,omitempty"`
	Role           string   `json:"role"`
	Members        []string `json:"members"`
	MaxConcurrency int      `json:"max_concurrency,omitempty"`
}

// SquadPutResp 是 PUT 成功响应：version 是本次写入产生的版本（expect+1）；
// 他人的并发写会再推进，编辑回路每次 GET 重取。
type SquadPutResp struct {
	Name    string `json:"name"`
	Version int    `json:"version"`
}

// QueueEntry 是 GET /api/queue 的元素。Kind/ID/Seq/Position 来自队列元数据，
// 其余九个字段逐一来自入队时刻的 IgnitionRequest 快照（拍板记录②：出队前
// 不重读卡）。Position 是清队顺序下的全局 1 基位次。
type QueueEntry struct {
	Kind     string `json:"kind"`
	ID       string `json:"id"`
	Card     string `json:"card"`
	Node     string `json:"node,omitempty"`
	Squad    string `json:"squad"`
	Target   string `json:"target,omitempty"`
	Executor string `json:"executor,omitempty"`
	Model    string `json:"model,omitempty"`
	Priority string `json:"priority,omitempty"`
	Ready    bool   `json:"ready"`
	Actor    string `json:"actor"`
	Seq      int64  `json:"seq"`
	Position int    `json:"position"`
}

// QueueResp 是 GET /api/queue 的响应体；空队列 = "queue":[]。
type QueueResp struct {
	Queue []QueueEntry `json:"queue"`
}

// CoordinatorLaunchResp 是 POST /api/cards/{id}/coordinator/launch 的成功响应
// （keystone.RoundResult 的 wire 投影，K4 的 handler 产出、K3 的 CLI 消费——
// 形状以此处为准，K4 不得另行发明字段）。
type CoordinatorLaunchResp struct {
	Woke      bool   `json:"woke"`
	SessionID string `json:"session_id,omitempty"`
	Rebuilt   bool   `json:"rebuilt"`
	Escalated bool   `json:"escalated"`
	Output    string `json:"output,omitempty"`
}

// CoordinatorAttachInfo 是协调者会话的定位三元组；Machine 允许为空串表示本机。
// Dir 与 Command 均由服务端定位器产生，客户端不得自行拼接或改写。
type CoordinatorAttachInfo struct {
	Machine string `json:"machine"`
	Dir     string `json:"dir"`
	Command string `json:"command"`
}

// CoordinatorStatus 是 GET coordinator 的状态；Attach=nil 序列化为 null 表示未绑定。
// attach_active 是进程内的人工接管态，与 Attach 三元组是否存在是两件事。
type CoordinatorStatus struct {
	Bound        bool                   `json:"bound"`
	AttachActive bool                   `json:"attach_active"`
	Attach       *CoordinatorAttachInfo `json:"attach"`
}

// CoordinatorAttachReleaseResp 是 active=false 交回无头后的成功回执。
type CoordinatorAttachReleaseResp struct {
	OK bool `json:"ok"`
}

// HomeProbeReq 是 POST /api/host/probe 的请求体。machine 走 query（?machine=），
// 不进 body——转发基座靠 query 选路。Credential 空 = standalone。
type HomeProbeReq struct {
	CLI        string `json:"cli"`
	Path       string `json:"path"`
	Credential string `json:"credential,omitempty"`
}

// HomeProbeResp 是路径探测的三类结果。kind 只允许 empty / logged_in / occupied。
type HomeProbeResp struct {
	Kind   string `json:"kind"`
	Detail string `json:"detail,omitempty"`
}

// HomeWakeReq 是 POST /api/host/wake 的请求体（本机唤起，供检测编排转发）。
type HomeWakeReq struct {
	CLI        string `json:"cli"`
	HomeDir    string `json:"home_dir"`
	Credential string `json:"credential,omitempty"`
	Model      string `json:"model,omitempty"`
}

// HomeWakeResp 是本机唤起结局。outcome 只允许 ready / need_login / quota / unreachable。
type HomeWakeResp struct {
	Outcome string `json:"outcome"`
	Detail  string `json:"detail,omitempty"`
}

// CarrierDetectResp 是 POST /api/squads/carriers/{name}/detect 成功后的投影。
type CarrierDetectResp struct {
	Name      string `json:"name"`
	Status    string `json:"status"`
	LastError string `json:"last_error,omitempty"`
	Version   int    `json:"version"`
}

// CarrierRunCommandResp 是 GET /api/squads/carriers/{name}/run-command 的回执。
// command 由服务端生成，客户端只复制。
type CarrierRunCommandResp struct {
	Command string `json:"command"`
}
